async/src/tests.rs

```diff
@@ -168,3 +169,52 @@ fn wait() {
         },
     );
 }
+
+#[test]
+fn drop_close_prod() {
+    let (prod, mut cons) = AsyncHeapRb::<usize>::new(1).split();
+    let stage = Arc::new(AtomicUsize::new(0));
+    let stage_clone = stage.clone();
+    let t0 = std::thread::spawn(move || {
+        execute!(async {
+            drop(prod);
+            assert_eq!(stage.fetch_add(1, Ordering::SeqCst), 0);
+        });
+    });
+    let t1 = std::thread::spawn(move || {
+        execute!(async {
+            cons.wait_occupied(1).await;
+            assert_eq!(stage_clone.fetch_add(1, Ordering::SeqCst), 1);
+            assert!(cons.is_closed());
+        });
+    });
```

Comment:
Must separate as two polling, otherwise waking the wrong party also passes this test.

Reply:
Good note, `join!`-ing multiple futures can actually lead to falsely passing this test because of spurious wake-ups.

Also the usage of some async runtime (like Tokio or async-std) may be considered for testing. Each `async` block then can be ran in separate task.